### PR TITLE
Generate pedigree

### PR DESCRIPTION
**Pull request by MattWellie**
# Fixes

  - N/A

## Proposed Changes

  - helper script for generating family data from sample-metadata:
    1. Pedigree
    2. Pedigree in PLINK format
    3. sample -> participant reverse lookup

## Checklist

- [ ] Related Issue created
- [x] Tests covering new change
- [x] Linting checks pass

## Misc

Early version of this script (#1) made some incorrect assumptions about the relationship between participants and samples. This version allows for a 1 -> n relationship, and all possible variations will be present in the final PED file

Use of the de novo functionality in Hail required the generation of a Pedigree in PLINK format, so that has been added as an option. I've confirmed that Peddy also accepts PLINK and .ped files interchangeably, so .fam is the new default for this application.

The mapping back from samples to participants is required when formatting results for presentation back to clinical analysts, so creation of that mapping has been included as a step here.

### CODE REVIEW

File: design_docs/Helpers.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+- export the Pedigree with all individuals as singletons
+
+### Pedigree manipulation
+
+The Pedigree data pulled directly from the SM-API contains all members and relationships in the context
+of the participants' external identifiers (PID). This can't be matched up with the VCF contents, as in
+sequence data files processing is instead done in the context of the Sample IDs (SID).
+
+To complicate things, there is no guarantee of a 1:1 relationship between participants and samples (e.g.
+replacement samples, but also Genome and Exome library preparations would be considered separate samples).
```
Comment (illusional) [inline design_docs/Helpers.md:30]: This assumption isn't true, most of the time genomes and exomes will have the same sample ID, but different sequence IDs.

File: design_docs/Helpers.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+To overcome any potential ambiguity, this process uses the following steps:
+
+- obtain the Pedigree data with external identifiers
+- obtain the mapping of PID:SID, allowing for multiple SID/PID
+- dump a dictionary to file, allowing lookup of all SID -> PID (guaranteed 1:1)
```
Comment (illusional) [inline design_docs/Helpers.md:35]: This sort of isn't correct, it's still `many samples:1 participant`, but I see your point is you can easily look it up.

File: design_docs/Helpers.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+- for each PID, collect all possible SID as a list
+- when writing the translated pedigree:
+  - use itertools.product to generate intra-family combinations of the proband, father, & mother e.g.
+  - 1 sample per participant == 1^3 =  row
```
Comment (illusional) [inline design_docs/Helpers.md:39]:
```suggestion
  - 1 sample per participant == 1^3 = row
```

File: helpers/pedigree_from_sample_metadata.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+
+def get_fat_pedigree(
+    pedigree_dicts: List[Dict[str, Union[str, List[str]]]],
+    sample_to_cpg_dict: Dict[str, List[str]],
+    singles: bool,
+    plink: bool,
+) -> List[Dict[str, List[str]]]:
+    """
+    swaps sample IDs for the internal CPG values
```
Comment (illusional) [inline helpers/pedigree_from_sample_metadata.py:33]: Is this participant IDs?
```suggestion
    swaps participant IDs for the internal CPG values
```

File: helpers/pedigree_from_sample_metadata.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    :param plink:
+    :return:
+    """
+
+    new_entries = []
+
+    for counter, ped_entry in enumerate(pedigree_dicts, 1):
+
+        if ped_entry['individual_id'] not in sample_to_cpg_dict:
+            continue
```
Comment (illusional) [inline helpers/pedigree_from_sample_metadata.py:49]: This is an error though right - or maybe the participant doesn't have any samples. I think a comment here might explain this case.
You might still want to proceed through, to show two people are related through a dummy individual

File: helpers/pedigree_from_sample_metadata.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    'maternal_id',
+    'sex',
+    'affected',
+]
+
+
+def get_fat_pedigree(
+    pedigree_dicts: List[Dict[str, Union[str, List[str]]]],
+    sample_to_cpg_dict: Dict[str, List[str]],
+    singles: bool,
```
Comment (illusional) [inline helpers/pedigree_from_sample_metadata.py:29]: Could this be described as an action, eg: `make_singletons`, and link it to the description in the docstring

File: helpers/pedigree_from_sample_metadata.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+    sample_map = {}
+    for (
+        participant,
+        sample,
+    ) in ParticipantApi().get_external_participant_id_to_internal_sample_id(
+        project=project
+    ):
+        sample_map.setdefault(participant, []).append(sample)
+    return sample_map
```
Comment (illusional) [inline helpers/pedigree_from_sample_metadata.py:138]: Using a defaultdict is another alternative

```suggestion
    sample_map = defaultdict(list)
    for (
        participant,
        sample,
    ) in ParticipantApi().get_external_participant_id_to_internal_sample_id(
        project=project
    ):
        sample_map[participant].append(sample)
    return sample_map
```

File: helpers/pedigree_from_sample_metadata.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    :param mapping_digest: created by ext_to_int_sample_map
+    :return:
+    """
+
+    reverse_lookup = {}
+    for participant, samples in mapping_digest.items():
+        for sample in samples:
+            reverse_lookup[sample] = participant
+
+    return reverse_lookup
```
Comment (illusional) [inline helpers/pedigree_from_sample_metadata.py:153]: Python dict comprehension is pretty good for this, though it is a big of a longer line

```suggestion
    return {
        sample: participant
        for participant, samples in mapping_digest.items()
        for sample in samples
    }
```

File: test/test_generate_pedigree.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+    :param get_ped_patch:
+    :return:
+    """
+
+    with open(JSON_PED, 'r', encoding='utf-8') as handle:
+        payload = json.load(handle)
+        get_ped_patch.return_value = payload
+        result = get_pedigree_for_project(project=PROJECT)
+        assert result == payload
```
Comment (illusional) [inline test/test_generate_pedigree.py:64]: I don't think this is a useful test, the inner function only makes one call that's patched

File: test/test_generate_pedigree.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+        payload = json.load(handle)
+        map_mock.return_value = payload
+        result = ext_to_int_sample_map(project=PROJECT)
+        assert isinstance(result, dict)
+        assert result == {
+            'FAM1_father': ['CPG11'],
+            'FAM1_mother': ['CPG12'],
+            'FAM1_proband': ['CPG13'],
+            'FAM2_proband': ['CPG41'],
+        }
```
Comment (illusional) [inline test/test_generate_pedigree.py:90]: It doesn't matter too much, but good to close the file once you're finished with it, then do stuff after file is closed.

```suggestion
    map_mock.return_value = payload
    result = ext_to_int_sample_map(project=PROJECT)
    assert isinstance(result, dict)
    assert result == {
        'FAM1_father': ['CPG11'],
        'FAM1_mother': ['CPG12'],
        'FAM1_proband': ['CPG13'],
        'FAM2_proband': ['CPG41'],
    }
```

File: test/test_generate_pedigree.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
+"""
```
Comment (illusional) [inline test/test_generate_pedigree.py:1]: Are these tests being run through CI? Or is that a later change once all the small WIP PRs have been merged?

**Review by illusional — CHANGES_REQUESTED** [review]
Looks great! A few minor changes :)

File: helpers/pedigree_from_sample_metadata.py (hunk 5 above)
Comment (illusional) [reply helpers/pedigree_from_sample_metadata.py:49]: This is the main actionable one for me :)

Comment (MattWellie) [reply helpers/pedigree_from_sample_metadata.py:49]: replaced this by catching all failures, and throwing an exception at the end of the iteration to say 'these samples came from one endpoint, but not the other'.

**Comment by MattWellie** [comment]
welp, rebased this after the recent merge to main and it's gone insane

**Comment by MattWellie** [comment]
Closing as superseded, suggested changes made in a separate (cleaner) merge

**Pull request closed without merging**